ch6 : data analysis - re-write (#308)
worldbank/d4di: chapters/data-analysis.tex

%------------------------------------------------

\begin{fullwidth}
Transforming raw data into a substantial contribution to scientific knowledge 
requires a mix of subject expertise, programming skills, 
and statistical and econometric knowledge. 
The process of data analysis is, therefore, 
a back-and-forth discussion between people 

> with differing experiences, perspectives, and research interests. 

Author: _a back-and-forth discussion between people with differing experiences, perspectives, and research interests._ I think _skill sets_ fits better in this list of characteristics than either _experiences_ or _perspectives_.

Reviewer: Changed to skill sets in 97b2ccd.

The research assistant usually ends up being the pivot of this discussion. 
It is their job to translate the data received from the field into
economically meaningful indicators and to analyze them 

> while making sure that code and outputs do not become tangled and lost over time.

Author: Two points related to _while making sure that code and outputs do not become tangled and lost over time._

* Is _tangled_ the best word?
* Shouldn't if be "_tangled **or** lost over time_"? It depends on what you want to say with tangled, which I am not 100% sure of

Reviewer: >while making sure that code and outputs do not become too difficult to follow or get lost over time.

(97b2ccd)

When it comes to code, though, analysis is the easy part, 

> as long as you have organized your data well. 

Author: can we emphasize this? otherwise I think this point might be overseen by most readers. I am fine just putting the last part in italics, like `\textit{as long as you have organized your data well}.`

Reviewer: Done in 392ac7c.

Of course, there is plenty of complexity behind it: 
the econometrics, the theory of change, the measurement methods, and so much more.
But none of those are the subject of this book. 

> Instead, this chapter will focus on how to organize your data work.

Author: Related to my point above about emphasizing _as long as you have organized your data well_. I think we can tie back to that point here. As in:

_Instead, this chapter will focus on how to organize your data work so that the coding the analysis becomes easy_.

Reviewer: Included your suggestion in c80a095.

The underlying data structure does not change.
The cleaned data set should contain only the variables collected in the field.
No modifications to data points are made at this stage, except for corrections of mistaken entries.

Cleaning is probably the most time consuming of the stages discussed in this chapter.
This is the time when you obtain an extensive understanding of  the contents and structure of the data that was collected.
Explore your data set using tabulations, summaries, and descriptive plots.
You should use this time to understand the types of responses collected, both within each survey question and across respondents.
Knowing your data set well will make it possible to do analysis.

% Deidentification ------------------------------------------------------------------
The initial input for data cleaning is the raw data.
It should contain only materials that are received directly from the field.
They will invariably come in a host of file formats and nearly always contain personally-identifying information.\index{personally-identifying information}
These files should be retained in the raw data folder \textit{exactly as they were received}.

> The folder must be encrypted if it is shared in an insecure fashion,\sidenote{\url{https://dimewiki.worldbank.org/wiki/Encryption}}

Author: All secure sharing includes encryption. So I think we should say that this folders must always be encrypted. And we can add that they need to make sure the software they use for sharing does indeed encrypt

Reviewer: I'm not sure I get it. In our case, for example, if a file is in one drive, we don't need to encrypt it, right? Is it because one drive already encrypts it?

Reviewer 2: I understand there are two ways encryption can happen.

Either the whole storage is encrypted at rest and permissioned/passworded/2FA'd (Bank-managed laptop; Bank-managed OneDrive), in which case the file does not need to be separately encrypted from the system; or

The storage is not that type (personal machine or personal storage like Dropbox), in which case the file itself does needs to be separately encrypted.

Do we broadly agree? For almost all practical applications this is going to be the second case...

Author: @luizaandrade , yes the only reason we do not need to encrypt it is that the WB OneDrive system does encrypt it appropriately for us.

@bbdaniels , yes, but there is more detail. DropBox is encrypted at rest but they own the key, and since DropBoc Inc is not on the IRB that is an issue. 

I think where I am sub-consciously taking this argument is that while encryption is binary (a file is either encrypted or not) a whole system is much more complex. If you share with DropBox the file is encrypted in transit and at rest, but the way it is encrypted at rest is not satisfactory.

Expressing this in a correct but still in a way helpful to the reader will take space in the book. So I guess the question is if we want to take up space for that.

Reviewer: 97b2ccd reads:

>Additionally, no one who is not listed in the IRB should be able to access its content, not even the company providing file-sharing services. Check if your organization has guidelines on how to store data securely, as they may offer an institutional solution. If that is not the case, you will need to encrypt the file and make sure that only IRB-listed team members have the encryption key.\sidenote{\url{https://dimewiki.worldbank.org/wiki/Encryption}}.

Author: I really like this. I just made a small edit in be5e8f020c44e4579594fff706d2c0d626f2ad7b. We seem to be working on this chapter at the same time, so I created branch for small edits like this that I will merge to `la-ch6-from-dev` as soon as I have read through this

and it must be backed up in a secure offsite location.

> Everything else can be replaced, but raw data cannot.

Author: I do not think _replaced_ is the right word here. What about _re-generated_? Or _re-created from the raw data_?

Reviewer: Edited the wording in c80a095. Let me know what you think.

Reviewer: Edited in c80a095:

>Every other file is created from the raw data, and therefore can be recreated.
>The exception, of course, is the raw data itself, so it should never be edited directly.

> Therefore, raw data should never be interacted with directly.

Author: _raw data should never be interacted with directly_ - never is a strong word as sometimes you need to interact with it. I think the point we are making is that our work flow should not be set up that our software interact with this file on a day-to-day basis.

Reviewer: Here's how it reads in c80a095:

>Every other file is created from the raw data, and therefore can be recreated.
>The exception, of course, is the raw data itself, so it should never be edited directly.

What do you think?

Reviewer 2: Maybe add:

> If it is absolutely necessary to edit the raw data by hand before processing, make a copy in the same location and perform the edits there; be sure to fully document them.

Reviewer: I don't think I see where you are going. I can't think of a case where someone would edit the raw data by hand.

Author: I have experienced that you need to remove some weird language character before you can import it without rows break incorrectly or something. 

I added some text on that here: 414f3b7dbdd0033636d254b0df0c959765b0b160 . I think that explains what I think where @bbdaniels were going, but I do not know if it is common enough that we even have to mention it. @luizaandrade, does the exception make sense at least?

Reviewer: ok, if the weird character is breaking rows, then I think it warrants manual edits.

Reviewer: made small changes to the wording in 7527ffa

> Secure storage of the raw\sidenote{\url{https://dimewiki.worldbank.org/wiki/Data\_Security}}

Author: Strange placement of this sidenote. Should this sidenote reference _Secure storage_ or the full sentence? Not the word _raw_ at least.

Reviewer: Moved to the end of the sentence in c80a095.

data means access to it will be restricted even inside the research team.

> Loading encrypted data multiple times it can be annoying.

Author: _Loading encrypted data multiple times it can be annoying._ A few things:
* there is a typo in _ it can be_
* multiple is not the right work here. should we say _frequently_ or _regularly_?
* Annoying is a negatively charged word. What about: _can be disruptive to the work flow_?

Reviewer: In c80a095:
> Loading encrypted data frequently can be disruptive to the workflow.

To facilitate the handling of the data, remove any personally identifiable information from the data set.
This will create a de-identified data set, that can be saved in a non-encrypted folder. 
De-identification,\sidenote{\url{https://dimewiki.worldbank.org/wiki/De-identification}}
at this stage, means stripping the data set of direct identifiers such as names, phone numbers, addresses, and geolocations.\sidenote{\url{ https://www.povertyactionlab.org/sites/default/files/resources/J-PAL-guide-to-deidentifying-data.pdf}}
The resulting de-identified data will be the underlying source for all cleaned and constructed data.
Because identifying information is typically only used during data collection, 
to find and confirm the identity of interviewees, 
de-identification should not affect the usability of the data.
In fact, most identifying information can be converted into non-identified variables for analysis purposes
(e.g. GPS coordinates can be translated into distances). 
However, if sensitive information is strictly needed for analysis,

> all the tasks described in this chapter must be performed in a secure environment.

Author: What is a _secure environment_? I have an idea, but we haven't explained that concept enough in this book to be able to say that here and not explain it further. What exactly is it you refer to here?

Reviewer 2: Hard to say -- this should be defined in IRB documentation and/or organizational guidance. In some cases it's sufficient to do on a company-managed laptop but sometimes you would need like a cold machine.

Reviewer: added examples in d09c32d

Author: I was mostly just pointing out that we kind of introduced a new concept _secure environment_ without explaining it. I think we are fine here if we just keep using the same terminology we have used before.

"_However, if sensitive information is strictly needed for analysis, then we need to keep our data encrypted while doing the tasks described in this chapter._"

(somehow the suggestion function was not working for me)

Reviewer: Changed in 7527ffa

% Unique ID and data entry corrections ---------------------------------------------
There are two main cases when the raw data will be modified during data cleaning.
The first one is when there are duplicated entries in the data.
Ensuring that observations are uniquely and fully identified\sidenote{\url{https://dimewiki.worldbank.org/wiki/ID\_Variable\_Properties}}
is possibly the most important step in data cleaning

> (as anyone who ever tried to merge data sets that are not uniquely identified knows).

Author: Merging a data set that is not fully unique is just a problem if you are doing a 1:1 merge. So this sentence needs to be more specific.

Reviewer 2: How about:

> The first one is when there are improperly duplicated entries in the data.
> Ensuring that observations are correctly and fully identified\sidenote{\url{https://dimewiki.worldbank.org/wiki/ID\_Variable\_Properties}}
> is possibly the most important step in data cleaning.

Author: Agreed. All you did was to remove the last sentence "_(as anyone who ever tried to merge data sets that are not uniquely identified knows)_", right?

Reviewer: done in d09c32d

Most of a Research Assistant's time is spent cleaning data and getting it into the right format. 
When the practices recommended here are adopted,
analyzing the data is as simple as using a command that is already implemented in a statistical software. 

\end{fullwidth}

%------------------------------------------------

\section{Data management}
The goal of data management is to organize the components of data work 
so it can traced back and revised without massive effort.
In our experience, there are four key elements to good data management: 
folder structure, task breakdown, master scripts, and version control. 
A good folder structure organizes files so that any material can be found when needed.
It reflects a task breakdown into steps with well-defined inputs, tasks, and outputs.
This breakdown is applied to code, data sets, and outputs.
A master script connects folder structure and code.
It is a one-file summary of your whole project.
Finally, version histories and backups enable the team 
to edit files without fear of losing information.
Smart use of version control also allows you to track 
how each edit affects other files in the project.

> % Task breakdown

Reviewer 3: Follow the order in the intro paragraph: folder structure, task breakdown, master scripts, version control (so move this section to after folder structure, or change the intro paragraph)

Reviewer 3: Related, labeled subsections would help organize content / guide the reader

Reviewer: Done in 17174e8

This makes it easier to track tasks across scripts, 
and avoids duplication of code that could lead to inconsistent results. 
For each stage, there should be a code folder and a corresponding data set. 
The names of codes, data sets and outputs for each stage should be consistent,
making clear how they relate to one another. 
So, for example, a script called \texttt{clean-section-1} would create
a data set called \texttt{cleaned-section-1}.

The division of a project in stages also helps the review workflow inside your team.
The code, data and outputs of each of these stages should go through at least one round of code review.
During the code review process, team members should read and run each other's codes.
Doing this at the end of each stage helps prevent the amount of work to be reviewed to become too overwhelming.
Code review is a common quality assurance practice among data scientists.
It helps to keep the level of the outputs high, and is also a great way to learn and improve your code.

% Folder structure
There are many schemes to organize research data. 
Our preferred scheme reflects the task breakdown just discussed.

> \index{data organization}

Reviewer: Changed this a paragraph a little bit in f744e09, but I think it's possible you were looking at an older version of the chapter?

files and folder structure that comprise a project.  
Having a master script eliminates the need for complex instructions to replicate results. 
Reading the master do-file should be enough for anyone unfamiliar with the project
to understand what are the main tasks, which scripts execute them,
and where different files can be found in the project folder. 
That is, it should contain all the information needed to interact with a project's data work.

% Version control
Finally, everything that can be version-controlled should be. 
Version control allows you to effectively track code edits,
including the addition and deletion of files. 
This way you can delete code you no longer need, 
and still recover it easily if you ever need to get back previous work.
Both analysis results and data sets will change with the code.
You should have each of them stored with the code that created it.
If you are writing code in Git/GitHub,
you can output plain text files such as \texttt{.tex} tables
and metadata saved in \texttt{.txt} or \texttt{.csv} to that directory.
Binary files that compile the tables,
as well as the complete data sets, on the other hand,

> should be stored in your team's shared folder. 

Reviewer: New issue: #330

Whenever data cleaning or data construction codes are edited,
use the master script to run all the code for your project.
Git will highlight the changes that were in data sets and results that they entail. 

%------------------------------------------------

> \section{Data cleaning}

Reviewer: Created subsections in 4c62e83

data means access to it will be restricted even inside the research team.
Loading encrypted data multiple times it can be annoying.
To facilitate the handling of the data, remove any personally identifiable information from the data set.
This will create a de-identified data set, that can be saved in a non-encrypted folder. 
De-identification,\sidenote{\url{https://dimewiki.worldbank.org/wiki/De-identification}}
at this stage, means stripping the data set of direct identifiers such as names, phone numbers, addresses, and geolocations.\sidenote{\url{ https://www.povertyactionlab.org/sites/default/files/resources/J-PAL-guide-to-deidentifying-data.pdf}}
The resulting de-identified data will be the underlying source for all cleaned and constructed data.
Because identifying information is typically only used during data collection, 
to find and confirm the identity of interviewees, 
de-identification should not affect the usability of the data.
In fact, most identifying information can be converted into non-identified variables for analysis purposes
(e.g. GPS coordinates can be translated into distances). 
However, if sensitive information is strictly needed for analysis,

> all the tasks described in this chapter must be performed in a secure environment.

Reviewer: New issue: #331

% Unique ID and data entry corrections ---------------------------------------------
There are two main cases when the raw data will be modified during data cleaning.
The first one is when there are duplicated entries in the data.
Ensuring that observations are uniquely and fully identified\sidenote{\url{https://dimewiki.worldbank.org/wiki/ID\_Variable\_Properties}}
is possibly the most important step in data cleaning
(as anyone who ever tried to merge data sets that are not uniquely identified knows).
Modern survey tools create unique observation identifiers.
That, however, is not the same as having a unique ID variable for each individual in the sample.
You want to make sure the data set has a unique ID variable
that can be cross-referenced with other records, such as the Master Data Set\sidenote{\url{https://dimewiki.worldbank.org/wiki/Master\_Data\_Set}}
and other rounds of data collection.
\texttt{ieduplicates} and \texttt{iecompdup}, 
two Stata commands included in the \texttt{iefieldkit} 
package\index{iefieldkit},\sidenote{\url{https://dimewiki.worldbank.org/wiki/iefieldkit}}
create an automated workflow to identify, correct and document

> occurrences of duplicate entries. 

Reviewer: New issue: #331

Looking for duplicated entries is usually part of data quality monitoring,
as is the only other reason to change the raw data during cleaning:
correcting mistakes in data entry.
During data quality monitoring, you will inevitably encounter data entry mistakes,
such as typos and inconsistent values.

> If you don't, you are probably not doing a very good job at looking for them.

Author: This sentence has a strange negative tone to it

Reviewer: ```suggestion
```

These mistakes should be fixed in the cleaned data set,
and you should keep a careful record of how they were identified,
and how the correct value was obtained.

% Data description ------------------------------------------------------------------
Note that if you are using secondary data, 

> the tasks described above can likely be skipped.

Author: _tasks described above can likely be skipped_ - I dont think they should ever be skipped. you have to ensure things like unique ID, look for mistakes etc, were addressed by the person that created and published this data. There are many mistakes in secondary data. So we should say something about that. what we can say tho is that you will most likely spend much less time on this with secondary data, but you should always check for this in any data you are about to use.

Reviewer: changed in 968ba4c

However, the last step of data cleaning, describing the data,

> will probably still be necessary.

Author: I don't know about using the word _describing_ here. I had to read a couple of times to understand myself. Maybe the sentence needs to be re-organized, as the first times I read this I kept reading it as if it was a list where _data cleaning_ was the first item, and _describing the data_ was the second item, not that the latter was a part of the first.

One quick fix would be _However, the last step of data cleaning — describing the data — will probably still be necessary.

Reviewer: Change in c80a095:
>Note that if you are using secondary data, the tasks described above can likely be skipped.
>The last step of data cleaning, however, will most likely still be necessary.
>It consists of describing the data, so that its users have all the information needed to interact with it.

Reviewer 2: "annotating"?

Reviewer: Changed to "annotate" in 968ba4c. I was trying to avoid using "documenting" so that people don't think that this is all that is needed for documentation. But maybe elaborating more is preferable?

Author: What about "labelling and annotating"? I think annotating is correct but it would have made my list of necessarily difficult words. And it is more than just labelling. So that is why I think both should be used, _labelling_ for clarity, _annotating_ to make sure it includes enough. I did that edit in b595afd67ce53c4c252eaf30159c43dcd843ae81. Feel free to revert if you do not like it.

Reviewer: hahaha. fair enough. that seems good to me. marking this conversation as resolved.

This is a key step to making the data easy to use, but it can be quite repetitive.
The \texttt{iecodebook} command suite, part of \texttt{iefieldkit},
is designed to make some of the most tedious components of this process,
such as renaming, relabeling, and value labeling,

> much easier (including in data appending).\sidenote{\url{https://dimewiki.worldbank.org/wiki/iecodebook}}

Author: _(including in data appending)_ - I know this is not incorrect, but it doesn't really fit the flow of the sentence. I think that this should be made it's own sentence and use a few more words saying that the process can be done slightly differently for different data sets, so that they can be appended. Or something like that.

Reviewer: Data appending is actually of construction, so I just removed the sentence in c80a095.

> such as renaming, relabeling, and value labeling,

Reviewer: addressed in 66e2227

\index{iecodebook}
We have a few recommendations on how to use this command for data cleaning.
First, we suggest keeping the same variable names as in the survey instrument,

> so it's easy to connect the two files.

Author: Is _connect_ a word we use for this? What about:
* _so it's easy to cross-reference the two files_
* _so it's easy to link the same variable across the two files_
* _so it's easy to compare the same variable across the two files_

Reviewer: In c80a095:

>First, we suggest keeping the same variable names as in the survey instrument, so it's straightforward to link data points for a variable to the question that originated them.

Author: There is a typo or some missing words:

_First, we suggest keeping the same variable names in **the cleaned data set** as in the survey instrument, so it's straightforward to link data points for a variable to the question that originated them._

Reviewer: Added "in the cleaned data set" in 97b2ccd.

i.e. per survey instrument.
Each row in the cleaned data set represents one survey entry or unit of observation.\sidenote{\cite{tidy-data}}
If the raw data set is very large, or the survey instrument is very complex,
you may want to break the data cleaning into sub-steps, 
and create intermediate cleaned data sets
(for example, one per survey module).
Breaking cleaned data sets into the smallest unit of observation inside a roster
make the cleaning faster and the data easier to handle during construction.
But having a single cleaned data set will help you with sharing and publishing the data.
To make sure this file doesn't get too big to be handled,
use commands such as \texttt{compress} in Stata to make sure the data
is always stored in the most efficient format.

% Documentation
Throughout the data cleaning process, you will need inputs from the field, 
including enumerator manuals, survey instruments, 
supervisor notes, and data quality monitoring reports.
These materials are part of what we call data documentation
\sidenote{\url{https://dimewiki.worldbank.org/wiki/Data\_Documentation}}
\index{Documentation},
and should be stored in the corresponding folder, 

> as you will probably need them during analysis and publication.

Reviewer 3: We should take a stronger position on this. These materials are essential for data documentation, and must be made available for publication.

Reviewer: changed the wording in b9cc5f1

Include in the \texttt{Documentation} folder records of any
corrections made to the data, including to duplicated entries,
as well as communications from the field where theses issues are reported.
Make sure to also have a record of potentially problematic patterns you noticed

> while exploring the data, such as outliers and variables with many missing values.

Reviewer 3: Do we have a best practice for how to store this kind of information? i've seen anything from comments in the do-file to a word doc with notes. it's never been obvious to me how to make sure that is fully linked back into the workflow either.

Reviewer: What I have done in the past is to create a tex file compiling plots and tabulates. But I think this is one case where markdown is really useful.

Reviewer: now in #346

Be very careful not to include sensitive information in 
documentation that is not securely stored, 
or that you intend to release as part of a replication package or data publication.

> 

Reviewer 3: the output of this stage is typically the publishable data, yes? if so, that's worth noting.

Reviewer: done in 3d73788

\section{Indicator construction}

% What is construction -------------------------------------

> Any changes to the original data set happen during construction.

Author: _changes_ is too vague here. It might be read as we make arbitrary changes to the original data. We should call it _corrections_ or something that makes it sound more formal.

Reviewer 2: > Data construction is the process of interpreting the data points as provided in the raw data to make them suitable for analysis.

Reviewer: Changed line 261 to Ben's suggestion in fe5b309

> Any changes to the original data set happen during construction.

Reviewer 3: Make the organization of the chapter clear. Change the first sentence to something like: the second step of data work is indicator construction: this is when you transform the raw dataset into the analysis dataset.

Reviewer 3: Also we should make reference here that indicator construction should be guided by the PAP

Reviewer: done in 3d73788

It is at this stage that the raw data is transformed into analysis data.
This is done by creating derived variables
(binaries, indices, and interactions, to name a few).
To understand why construction is necessary,
let's take the example of a household survey's consumption module.
It will result in separate variables indicating the 
amount of each item in the bundle that was consumed.
There may be variables indicating the cost of these items.
You cannot run a meaningful regression on these variables. 

> You need to manipulate them into something that has \textit{economic} meaning. 

Author: This example is a bit confusing. I am all in favour of specific examples to help with intuition, but this is half-specific as we make the specific example of a consumption module, and make non specific comments of what goes into a consumption module. Let me know if you want my help to re-think this example.

Reviewer: how about 092fe91?

During this process, the data points will typically be reshaped and aggregated 
so that level of the data set goes from the unit of observation in the survey to the unit of analysis.\sidenote{\url{https://dimewiki.worldbank.org/wiki/Unit\_of\_Observation}} 
To use the same example, the data on quantity consumed was collect for each item, and needs to be aggregated to the household level before analysis.

% Why it is a separate process -------------------------------

% From cleaning
Construction is done separately from data cleaning for two reasons. 
The first one is to clear differentiation between the data originally collected and the result of data processing decisions.
The second is to ensure that variable definition is consistent across data sources. 
Unlike cleaning, construction can create many outputs from many inputs. 
Let's take the example of a project that has a baseline and an endline survey. 

> Unless the two instruments are exactly the same, which is unlikely,  the data cleaning for them will require different steps, and therefore will be done separately. 

Author: _, which is unlikely,_ - it sounds as if we are indifferent to whether it is or not. how about:

_, which is preferably but often not the case,_

Reviewer: >Unless the two instruments are exactly the same, which is preferable but often not the case,  the data cleaning for them will require different steps, and therefore will be done separately.

(c80a095)

Reviewer 3: Even if the instruments are the same, data cleaning will be done at different times and should be different - the same errors are not going to arise even if exactly the same instrument is fielded twice, and there's also sample changes to consider (e.g. attrition)

However, you still want the constructed variables to be calculated in the same way, so they are comparable.

> So you want to construct indicators for both rounds in the same code, after merging them.

Reviewer 3: "So, you will clean the data for the two survey rounds separately, but then use one script to construct the indicators for both rounds, so that they are exactly comparable for analysis purposes"

However, i'm not sure this is practical - often we do construction after each round. would the recommendation then be at the end of the project to merge the indicator construction into one? or this only applies to data work after a project is completed?

Reviewer: Addressed in 0602589 and daf5e91

% From analysis
Data construction is never a finished process.
It comes ``before'' data analysis only in a limited sense: the construction code must be run before the analysis code.
Typically, however, construction and analysis code are written concurrently.

> As you write the analysis, different constructed variables will become necessary, as well as subsets and other alterations to the data.

Reviewer 3: *Ideally* indicator construction can be done first (by following a PAP) - but in practice it's almost never that neat

Reviewer: Phrased it differently in 17174e8

Still, constructing variables in a separate script from the analysis will help you ensure consistency across different outputs. 
If every script that creates a table starts by loading a data set, subsetting it and manipulating variables, any edits to construction need to be replicated in all scripts. 
This increases the chances that at least one of them will have a different sample or variable definition.

% What to do during construction -----------------------------------------

> Construction is the step where you face the largest risk of making a mistake that will affect your results. 

Author: _Construction is the step where you face the largest risk of making a mistake that will affect your results._ - source or justifcation of this statement?

Reviewer 2: Suggest to replace with:
> Because data construction involves translating concrete data points to more abstract measurements, it is important to document exactly how each is derived or calculated.

Reviewer: addressed in 5134e94

> Keep in mind that details and scales matter. 

Author: Is _scales_ the right word here? We mean units of measurement right? I think this can be removed from here as it is discussed properly later. Unless we want to make units/scales an example. This sentence now reads as if we want the reader to keep two things in mind, _details_ and _scales_, but what I think we mean is that scales is an example of a detail the reader should keep in mind.

Reviewer: ```suggestion
Keep in mind that details matter. 
```

Reviewer: addressed in 5134e94

It is important to check and double-check the value-assignments of questions and their scales before constructing new variables using them. 
Are they in percentages or proportions? 
Are all variables you are combining into an index or average using the same scale? 

> Are yes or no questions coded as 0 and 1, or 1 and 2?

Author: _Are yes or no questions **consistently** coded as 0 and 1, or 1 and 2?_

Reviewer 2: What about "binary" questions rather than yes/no? (Thinking also of gender-type stuff where I've seen very basic mistakes get made)

Reviewer: Ok, I see that this is not clear. But not sure how to address it. It doesn't really matter in this example if they are consistently coded, or if they are categorical or binary. It just matters that you're mindful of the underlying value. Will try to rephrase it.

Reviewer: ```suggestion
In Stata, categorical variables have underlying numerical values, so it is important to be mindful of what they are when using this type of variable: a variable with "Yes" and "No" options, for example, may be coded as 0 and 1, or just as well as 1 and 2.
```

Author: While possible to code differently, isn't very error prone? Should we recommend that all yes/no is coded into 1/0 so they numerically can be used as frequencies in means and dummies in regressions?

Reviewer: Yes, you're right. We should.

Reviewer 2: Yes we should but we should leave this part as is, so something like:

> Are yes/no questions coded as 0 and 1, or 1 and 2? Make sure they are recoded to 0 and 1. Similarly, you should typically generate dummies for each category in multiple-response variables (such as gender).

Author: Moved to issue #345

At this point, you will also need to address some of the issues in the data that you identified during data cleaning. 
The most common of them is the presence of outliers.
How to treat outliers is a research question, but make sure to note what we the decision made by the research team, and how you came to it. 
Results can be sensitive to the treatment of outliers, so keeping the original variable in the data set will allow you to test how much it affects the estimates.
More generally, create derived measures in new variables instead of overwriting the original information.

% Outputs -----------------------------------------------------------------

% Data set
The outputs of construction are the data sets that will be used for analysis.
The level of observation of a constructed data set is the unit analysis. 
Each data set is purpose-built to answer an analysis question.
Since different pieces of analysis may require different samples,
or even different units of observation,
you may have one or multiple constructed data sets, 

> depending on how your analysis is structured.

Author: _Since different pieces of analysis may require different samples, or even different units of observation, you may have one or multiple constructed data sets, depending on how your analysis is structured._ 

I think we need to explain this better to someone who has never though of this. As in how does it depend on how the analysis is structured? How many data sets if multiple? I think we are explaining an advanced workflow in quite a vague way

Reviewer 2: "Build the datasets you need, and not the ones you don't" ;-P

I think the key idea is that one project, or even one paper, is not one dataset. I'm not sure how to get that across exactly...

Reviewer: addressed in 342f6b8

So don't worry if you cannot create a single, ``canonical'' analysis data set.
It is common to have many purpose-built analysis datasets:
there may be a \texttt{data-wide.dta},
\texttt{data-wide-children-only.dta}, \texttt{data-long.dta},
\texttt{data-long-counterfactual.dta}, and many more as needed.
One thing all constructed data sets should have in common, though,
are functionally-named variables.
As you no longer need to worry about keeping variable names

> consistent with the survey, they should be as intuitive as possible.

Author: _As you no longer need to worry about keeping variable names consistent with the survey, they should be as intuitive as possible._ - We write _no longer worry_ as if it doesn't matter. It is great if they are when it can be combined, but it is more important that it is intuitive then matching, so it is not necessary that they are the same anymore, but we should not write this as if it doesn't matter at all.

Reviewer: >One thing all constructed data sets should have in common, though, are functionally-named variables.
>Constructed variables are called ``constructed'' because they were not present in the survey to start with,
so making their names consistent with the survey form is not as crucial.
>Of course, whenever possible, having variables names that are both intuitive and can be linked to the survey is ideal.
>However, functionality should be prioritized here.

(c80a095)

> consistent with the survey, they should be as intuitive as possible.

Reviewer 3: Mention earlier in this section that at this stage variable names should be functional, differentiating it from the first stage

Remember to consider keeping related variables together and adding notes to each as necessary.

% Documentation

> It is wise to start an explanatory guide as soon as you start making changes to the data.

Reviewer 3: what does an explanatory guide look like here? is that just comments in code? or a separate doc?

Reviewer: more detailed version in 3030909.

Carefully record how specific variables have been combined, recoded, and scaled. 
This can be part of a wider discussion with your team about creating protocols for variable definition.
That will guarantee that indicators are defined consistently across projects.

> Documentation is an output of construction as relevant as the codes.

Author: "_as relevant as code **and data**_"?

Reviewer: Changed to "the code and the data" in 97b2ccd.

and \textit{Causal Inference: The Mixtape}.\sidenote{\url{http://scunning.com/mixtape.html}}
This section will not include instructions on how to conduct specific analyses.
That is a research question, and requires expertise beyond the scope of this book.
Instead, we will outline the structure of writing analysis code,
assuming you have completed the process of data cleaning and construction.

% Exploratory and final data analysis -----------------------------------------
The analysis stage usually starts with a process we call exploratory data analysis.
This is when you are trying different things and looking for patterns in your data. 
It progresses into final analysis when your team starts to decide what are the main results, those that will make it into the research output.
The way you deal with code and outputs for exploratory and final analysis is different, and this section will discuss how.
% Organizing scripts ---------------------------------------------------------
During exploratory data analysis, you will be tempted to write lots of analysis into one big, impressive, start-to-finish script. 

> Though it's fine to write such a script during a long analysis meeting, this practice is error-prone.  

Author: "_during a long analysis meeting_"  - we need to explain what we mean here if this is something we want to keep

Reviewer: >Although it's fine to write such a script if you are coding in real-time during a long analysis meeting with your PIs, this practice is error-prone.  

(c80a095)

Author: I am not sure I get it still. Is this common that people write code in meetings? What is the point we are making here? That things does not have to be coded in the order they were discussed?

Then I also think it is a too explicit RA focus. I think we can say "_during a long analysis meeting with **the rest of your research team**_"

Reviewer: I certainly do. But I am not very attached to this sentence.

> It subtly encourages poor practices such as not clearing the workspace and not loading fresh data. 

Author: "_not loading fresh data **between each analysis task**_"?

Reviewer: >It subtly encourages poor practices such as not clearing the workspace and not loading fresh data for each analysis task. 

(c80a095)

Author: _It subtly encourages poor practices such as not clearing the workspace and not **re-loading the constructed data before** each analysis task._

The word _fresh_ here is a bit odd

Reviewer: >It subtly encourages poor practices such as not clearing the workspace and not reloading the constructed data set before each analysis task. 

(97b2ccd)

It's important to take the time to organize scripts in a clean manner and to avoid mistakes.

A well-organized analysis script starts with a completely fresh workspace and explicitly loads data before analyzing it.
This encourages data manipulation to be done earlier in the workflow (that is, during construction).

> It also and prevents you from accidentally writing pieces of analysis code that depend on one another, leading to the too-familiar ``run this part, then that part, then this part'' process.

Author: I understand what you mean by "_too-familiar ``run this part, then that part, then this part'' process_" but I do not think we can use this language here. What about:

_It also and prevents you from accidentally writing pieces of analysis code that depend on one another and requires manual instructions  for all required code snippets be run in the right order_

Reviewer: Adopted you suggestion in c80a095.

Each script should run completely independently of all other code.
You can go as far as coding every output in a separate script.
There is nothing wrong with code files being short and simple -- as long as they directly correspond to specific pieces of analysis.

> Analysis files should be as simple as possible, so you can focus on the econometrics.

Author: Who is _you_ in this sentence? I can guess but are not exactly sure which recommendation to which benefit we are giving here.

Reviewer: >Analysis files should be as simple as possible, so whoever is reading it can focus on the econometrics.

(c80a095)

All research decisions should be made very explicit in the code.
This includes clustering, sampling, and control variables, to name a few. 
If you have multiple analysis data sets, each of them should have a descriptive name about its sample and unit of observation.
As your team comes to a decision about model specification, you can create globals or objects in the master script to use across scripts.

> This is a good way to make sure specifications are consistent throughout the analysis. It's also very dynamic, making it easy to update all scripts if needed.

Author: What is `It's` in "_It's also very dynamic_"? Globals in the master script? After reading it many times I see what you are saying but it is a bit vague or imprecise.

Reviewer: >This is a good way to make sure specifications are consistent throughout the analysis. 
>Using pre-specified globals or objects also makes your code more dynamic, so it is easy to update specifications and results without changing every script.

(97b2ccd)

It is completely acceptable to have folders for each task, and compartmentalize each analysis as much as needed.

> It is always better to have more code files open than to keep scrolling inside a given file.

Author: "_It is always better **to have** to have more code files open than to keep scrolling inside a given file._" or "_It is always better **to be required** to have more code files open than to keep scrolling inside a given file._" Just having more code files open is not the trade-off we are talking about here. Does that make sense?

Reviewer: Not sure I get it. Let's talk about in the meeting.

Reviewer: I actually think we should scrape this sentence altogether. You shouldn't need multiple files open at once if they are independent of each other.
```suggestion
```

Reviewer: done in 13c1f85

To accomplish this, you will need to make sure that you have an effective data management system, including naming, file organization, and version control.
Just like you did with each of the analysis datasets, name each of the individual analysis files descriptively.

> Code files such as \path{spatial-diff-in-diff.do}, \path{matching-villages.do}, and \path{summary-statistics.do} 

Author: Very Stata specific examples? Can we make them more agnostic or include other less Stata specific examples?

Reviewer 2: lol sure change one to `.r` then ;-)

Reviewer: .R and .py in 83a294a

are clear indicators of what each file is doing, and allow you to find code quickly.
If you intend to numerically order the code as they appear in a paper or report, 
leave this to near publication time.

% Self-promotion ------------------------------------------------
Our team has created a few products to automate common outputs and save you 
precious research time.
The \texttt{ietoolkit} package includes two commands to export nicely formatted tables.

> \texttt{iebaltab}\sidenote{\url{https://dimewiki.worldbank.org/wiki/Iebaltab}} creates and exports balance tables to excel and {\TeX}. 

Author: We have consistently used LaTeX before. Why only TeX here?

Reviewer: Changed to LaTeX in c80a095. However, we should look up the difference between the two and be more precise in their usage.

Author: Ok, created issue in https://github.com/worldbank/d4di/issues/323. @luizaandrade , feel free to resolve this thread once you've seen this comment.

\texttt{ieddtab}\sidenote{\url{https://dimewiki.worldbank.org/wiki/Ieddtab}} does the same for difference-in-differences regressions.
The \textbf{Stata Visual Library}\sidenote{	\url{https://worldbank.github.io/Stata-IE-Visual-Library/}}
has examples of graphs created in Stata and curated by us.\sidenote{A similar resource for R is \textit{The R Graph Gallery}. \\\url{https://www.r-graph-gallery.com/}}
\textbf{Data visualization}\sidenote{\url{https://dimewiki.worldbank.org/wiki/Data\_visualization}} \index{data visualization}

> is increasingly popular, but a great deal lacks in quality.\cite{healy2018data,wilke2019fundamentals}

Author: This sentence is odd. What are we trying to say exactly?

Reviewer 2: people make crap graphs

Author: I got the last part of the sentence. But isn't "_Data visualization is increasingly popular_" a weird statement? I agree with what I understand that you are trying to say, but it seems like something one say in passing and not writing in a book. And data visualization has been very poplar for a long time, but there is a boom in it and it is turning into a field of its own. No strong feelings about this if you all think this is good.

Reviewer: I don't have strong feelings about this sentence, for one side or the other. I also don't have a better suggestion of phrasing, but we can brainstorm about it during the meeting.

Reviewer: removed in 3a9f1d8

For an easier way around it, Gray Kimbrough's \textit{Uncluttered Stata Graphs} code is an excellent default replacement for Stata graphics that is easy to install.
\sidenote{\url{https://graykimbrough.github.io/uncluttered-stata-graphs/}}
If you are a R user, the \textit{R Graphics Cookbook}\sidenote{\url{https://r-graphics.org/}} 
is a great resource for the most popular visualization package \texttt{ggplot}\sidenote{\url{https://ggplot2.tidyverse.org/}}. 
But there are a variety of other visualization packages, 
such as \texttt{highcharter}\sidenote{\url{http://jkunst.com/highcharter/}}, 
\texttt{r2d3}\sidenote{\url{https://rstudio.github.io/r2d3/}}, 
\texttt{leaflet}\sidenote{\url{https://rstudio.github.io/leaflet/}}, 
and \texttt{plotly}\sidenote{\url{https://plot.ly/r/}}, to name a few.
We have no intention of creating an exhaustive list, and this one is certainly missing very good references.
But at least it is a place to start.

\section{Exporting analysis outputs}

% Exploratory analysis
It's ok to not export each and every table and graph created during exploratory analysis. 

> Instead, we suggest running them into markdown files using RMarkdown or the different dynamic document options available in Stata. 

Author: sidenote to RMarkdown or explain it as we haven't mentioned it yet in this chapter

Reviewer 2: I'd just remove the sentence

Reviewer: Removed in 97166a9. It seems I have a very particular exploratory analysis workflow in mind.

This will allow you to update and present results quickly while maintaining a record of the different analysis tried. 
% Final analysis
Final analysis scripts, on the other hand, should export final outputs, which are ready to be included to a paper or report.

> No manual edits, including formatting, should be necessary after exporting final outputs. 

Author: You later get to copy and paste, but I think we already here should differentiate between different types of manual edits. Manual edits that reuire copy and paste is an absolute no-no with extremely few if any exceptions. But manual formatting is something that really should be avoided but is not as risky and copy-paste.

Reviewer: addressed in ab0718b

Author: I like it, just made one edit in 1d1d14a58dda41170d1865d8df6f2b6a645d2918. I changed _risk-prone_ to _error-prone_. @luizaandrade , feel free to resolve this comment if you are ok with that edit

There are numerous commands to export outputs from both R and Stata.\sidenote{Some examples are \href{ http://repec.sowi.unibe.ch/stata/estout/}{\texttt{estout}}, \href{https://www.princeton.edu/~otorres/Outreg2.pdf}{\texttt{outreg2}}, 
and \href{https://www.benjaminbdaniels.com/stata-code/outwrite/}{\texttt{outwrite}} in Stata, 
and \href{https://cran.r-project.org/web/packages/stargazer/vignettes/stargazer.pdf}{\texttt{stargazer}}
and \href{https://ggplot2.tidyverse.org/reference/ggsave.html}{\texttt{ggsave}} in R.}
Save outputs in accessible and, whenever possible, lightweight formats.
Accessible means that it's easy for other people to open them.
In Stata, that would mean always using \texttt{graph export} to save images as \texttt{.jpg}, \texttt{.png}, \texttt{.pdf}, etc., 
instead of \texttt{graph save}, which creates a \texttt{.gph} file that can only be opened through a Stata installation.
Some publications require ``lossless'' TIFF of EPS files, which are created by specifying the desired extension.
For tables, \texttt{.tex} is preferred. 
Excel \texttt{.xlsx} and \texttt{.csv} files are also acceptable, 

> although if you are working on a large report they will become cumbersome to update after revisions.

Author: We can be more specifc it here. It is _not_ difficult to update .xlsx and .csv files. You just re-run the code. But what is difficult is to update those files and make sure the report is updated.

Reviewer: > Excel \texttt{.xlsx} and \texttt{.csv} files are also acceptable, but require the extra step of copying the tables into the final output, so it can be cumbersome to ensure that your paper or report is always up-to-date.

(c80a095)

Reviewer 3: And the workflow that we want to discourage for sure is exporting tables to excel and then copy-pasting those into formatted tables in word. that's common and super error-prone!

Reviewer 2: > And the workflow that we want to discourage for sure is exporting tables to excel and then copy-pasting those into formatted tables in word. that's common and super error-prone!

Feeling seen 🥰

Reviewer: edited a bit in f6c9917

Whichever format you decide to use, remember to always specify the file extension explicitly.

% Formatting
If you need to create a table with a very particular format, that is not automated by any command you know, consider writing the it manually 
(Stata's \texttt{filewrite}, for example, allows you to do that).
This will allow you to write a cleaner script that focuses on the econometrics, and not on complicated commands to create and append intermediate matrices.
To avoid cluttering your scripts with formatting and ensure that formatting is consistent across outputs,
define formatting options in an R object or a Stata global and call them when needed.
% Output content
Keep in mind that final outputs should be self-standing.
This means it should be easy to read and understand them with only the information they contain.
Make sure labels and notes cover all relevant information, such as sample, unit of observation, unit of measurement and variable definition.\sidenote{\url{https://dimewiki.worldbank.org/wiki/Checklist:\_Reviewing\_Graphs} \\ \url{https://dimewiki.worldbank.org/wiki/Checklist:\_Submit\_Table}}

> 

Reviewer 3: Let's have a closing sentence or two that links to the next chapter, something along the lines of once data analysis is complete you're ready to publish (but hopefully better than that!)

Reviewer: First attempt in 529aa73
